src/rust/engine/fs/store/src/snapshot_ops.rs

```diff
@@ -406,6 +406,12 @@ impl IntermediateGlobbedFilesAndDirectories {
           RestrictedPathGlob::Wildcard { wildcard } => {
             // NB: We interpret globs such that the *only* way to have a glob match the contents of
             // a whole directory is to end in '/**' or '/**/*'.
+
+            if exclude.maybe_is_parent_of_ignored_path(&directory_path) {
+              // leave this directory in todo_directories, so we process ignore patterns correctly..
+              continue;
+            }
+
             if *wildcard == *DOUBLE_STAR_GLOB {
               globbed_directories.insert(directory_path, directory_node.clone());
             } else if !globbed_directories.contains_key(&directory_path) {
```

Comment:
I can't really make sense if what is going on here. As it seems to entirely side step any ignore patterns that may be in effect. Also, I was hitting the else here, adding an empty directory. But that directory was never populated with files, and I haven't figured out yet, when and where that was supposed to happen.

So, by simply not doing any of this, my tests passed, as the directories where processed in full.. maybe this is just an optimization?

Reply:
Yea, the skipping recursing into subdirectories based on wildcards is optimization. To the extent that it is buggy, we should bias toward correctness.

I had hoped that we would be able to use the same code here as we use for [glob matching against a filesystem](https://github.com/pantsbuild/pants/blob/2fbd8bb618cbc06bb242e461be9e9942e7699e65/src/rust/engine/fs/src/glob_matching.rs#L370-L380), but the primary blocking issue for that is exactly that this code is trying to shortcircuit and skip recursing into things (i.e. "match an entire subdirectory"), and that's not an operation that we can do in the filesystem (due to symlinks, etc).

It's possible that the implementations could be merged, but in the meantime: removing some optimization to ensure that this is correct is the only course.